diff --git a/lib/publish.js b/lib/publish.js
index ff3af00789942572806e52627b5e22c8ac25d1b1..8b1af5508a9b51874be1855cdde3fc551e93846f 100644
--- a/lib/publish.js
+++ b/lib/publish.js
@@ -4,7 +4,7 @@ const getRegistry = require('./get-registry');
 const getChannel = require('./get-channel');
 const getReleaseInfo = require('./get-release-info');
 
-module.exports = async (npmrc, {npmPublish, pkgRoot}, pkg, context) => {
+module.exports = async (npmrc, {npmPublish, pkgRoot, pkgManager}, pkg, context) => {
   const {
     cwd,
     env,
@@ -18,16 +18,30 @@ module.exports = async (npmrc, {npmPublish, pkgRoot}, pkg, context) => {
     const basePath = pkgRoot ? path.resolve(cwd, pkgRoot) : cwd;
     const registry = getRegistry(pkg, context);
     const distTag = getChannel(channel);
+    let command;
+    let args = ['publish', basePath, '--tag', distTag, '--registry', registry];
+    let userconfigOld;
+
+    if (pkgManager === 'pnpm') {
+      command = 'pnpm';
+      userconfigOld = process.env.npm_config_userconfig;
+      process.env.npm_config_userconfig = npmrc;
+    } else {
+      command = 'npm';
+      args.push('--userconfig', npmrc);
+    }
 
     logger.log(`Publishing version ${version} to npm registry on dist-tag ${distTag}`);
-    const result = execa(
-      'npm',
-      ['publish', basePath, '--userconfig', npmrc, '--tag', distTag, '--registry', registry],
-      {cwd, env, preferLocal: true}
-    );
+    const result = execa(command, args, {cwd, env, preferLocal: true});
     result.stdout.pipe(stdout, {end: false});
     result.stderr.pipe(stderr, {end: false});
-    await result;
+    try {
+      await result;
+    } finally {
+      if (pkgManager === 'pnpm') {
+        process.env.npm_config_userconfig = userconfigOld;
+      }
+    }
 
     logger.log(`Published ${pkg.name}@${version} to dist-tag @${distTag} on ${registry}`);
 