diff --git a/lib/publish.js b/lib/publish.js
index ac3e6feea8e388d88e77d58538a265f7ab4eac05..1feb6e3abe3e16f9a4fb3933128151c065782815 100644
--- a/lib/publish.js
+++ b/lib/publish.js
@@ -18,16 +18,30 @@ export default async function (npmrc, { npmPublish, pkgRoot }, pkg, context) {
     const basePath = pkgRoot ? path.resolve(cwd, pkgRoot) : cwd;
     const registry = getRegistry(pkg, context);
     const distTag = getChannel(channel);
+    let command;
+    let args = ['publish', basePath, '--tag', distTag, '--registry', registry];
+    let userconfigOld;
 
-    logger.log(`Publishing version ${version} to npm registry on dist-tag ${distTag}`);
-    const result = execa(
-      "npm",
-      ["publish", basePath, "--userconfig", npmrc, "--tag", distTag, "--registry", registry],
-      { cwd, env, preferLocal: true }
-    );
+    if (pkgManager === 'pnpm') {
+      command = 'pnpm';
+      userconfigOld = process.env.npm_config_userconfig;
+      process.env.npm_config_userconfig = npmrc;
+    } else {
+      command = 'npm';
+      args.push('--userconfig', npmrc);
+    }
+
+    logger.log(`Publishing version ${version} using ${command} to npm registry on dist-tag ${distTag}`);
+    const result = execa(command, args, {cwd, env, preferLocal: true});
     result.stdout.pipe(stdout, { end: false });
     result.stderr.pipe(stderr, { end: false });
-    await result;
+    try {
+      await result;
+    } finally {
+      if (pkgManager === 'pnpm') {
+        process.env.npm_config_userconfig = userconfigOld;
+      }
+    }
 
     logger.log(`Published ${pkg.name}@${version} to dist-tag @${distTag} on ${registry}`);
 
